stdlib/universe/triple_exponential_derivative.go

```diff
@@ -0,0 +1,373 @@
+
+		n: spec.N,
+	}
+}
+
+func (t *tripleExponentialDerivativeTransformation) RetractTable(id execute.DatasetID, key flux.GroupKey) error {
+	return t.d.RetractTable(key)
+}
+
+func (t *tripleExponentialDerivativeTransformation) Process(id execute.DatasetID, tbl flux.Table) error {
+	builder, created := t.cache.TableBuilder(tbl.Key())
+	if !created {
+		return fmt.Errorf("triple exponential derivative found duplicate table with key: %v", tbl.Key())
+	}
+	if t.n <= 0 {
+		return fmt.Errorf("cannot take triple exponential derivative with a period of %v (must be greater than 0)", t.n)
+	}
+	cols := tbl.Cols()
+	valueIdx := -1
+	for j, c := range cols {
+		if c.Label == execute.DefaultValueColLabel {
+			if c.Type != flux.TInt && c.Type != flux.TUInt && c.Type != flux.TFloat {
+				return fmt.Errorf("cannot take exponential moving average of column %s (type %s)", c.Label, c.Type.String())
+			}
+			valueIdx = j
+			mac := c
+			mac.Type = flux.TFloat
+			_, err := builder.AddCol(mac)
+			if err != nil {
+				return err
+			}
+		} else {
+			_, err := builder.AddCol(c)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	if valueIdx == -1 {
+		return fmt.Errorf("cannot find _value column")
+	}
+
+	// Keeps track of current position and last value looked at for columns that are passed through
+	// Faster than calling ema.PassThrough three times
+	t.i = make([]int, len(cols))
+	t.lastVal = make([]interface{}, len(cols))
+
+	t.ema1 = moving_average.New(int(t.n), len(cols))
+	t.ema2 = moving_average.New(int(t.n), len(cols))
+	t.ema3 = moving_average.New(int(t.n), len(cols))
+
+	err := tbl.Do(func(cr flux.ColReader) error {
+		if cr.Len() == 0 {
+			return nil
+		}
+
+		for j, c := range cr.Cols() {
+			isValueCol := false
+			if valueIdx == j {
+				isValueCol = true
+			}
+			var err error
+			switch c.Type {
+			case flux.TBool:
+				// We can pass through values using one of the EMAs, since the same number of values have to be appended
+				err = t.passThrough(moving_average.NewArrayContainer(cr.Bools(j)), builder, j)
+			case flux.TInt:
+				err = t.doFirstEMA(moving_average.NewArrayContainer(cr.Ints(j)), builder, j, isValueCol)
+			case flux.TUInt:
+				err = t.doFirstEMA(moving_average.NewArrayContainer(cr.UInts(j)), builder, j, isValueCol)
+			case flux.TFloat:
+				err = t.doFirstEMA(moving_average.NewArrayContainer(cr.Floats(j)), builder, j, isValueCol)
+			case flux.TString:
+				err = t.passThrough(moving_average.NewArrayContainer(cr.Strings(j)), builder, j)
+			case flux.TTime:
+				err = t.passThroughTime(cr.Times(j), builder, j)
+			}
+
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+
+	for j := range cols {
+		if j == valueIdx {
+			if err := t.doRest(builder, j); err != nil {
+				return err
+			}
+		} else {
+			// Check for any incomplete TRIXs, append the last value
+			if int64(t.i[j]) < 3*(t.n-1) {
+				val := values.New(t.lastVal[j])
+				if err := builder.AppendValue(j, val); err != nil {
+					return err
+				}
```

Comment:
Just for clarifying, does this happen only when you have less values than needed?

Anyways, why don't you simply store the last value as `vs.Value(vs.Len() - 1)` (with the appropriate null checks)?

Reply:
It seems to me that you can apply simple conditional logic to get the last value for a column (also using `3*(t.n-1)`) and use arrow slicing when passing through. If you need more details, please ask.